requestStorageAccessForOrigin.bs

```diff
@@ -0,0 +1,228 @@
+1. Let |existing state| be |descriptor|'s [=permission state=] with |settings|.
+1. If |existing state| is [=permission/granted=], return true.
+1. Return false.
+
+</div>
+
+<div algorithm>
+To <dfn type="abstract-op">determine the top-level storage access policy</dfn> for {{TopLevelStorageAccessPermissionDescriptor}} |descriptor|, with {{Document}} |doc|, [=site=] |top-level site|, and {{Promise}} |p|, run these steps:
+1. Let |settings| be |doc|'s [=relevant settings object=].
+1. Let |global| be |doc|'s [=relevant global object=].
+1. Let |existing state| be |descriptor|'s [=permission state=] with |settings|.
+1. If |existing state| is [=permission/granted=]:
+    1. [=Queue a global task=] on the [=permission task source=] given |global| to [=resolve=] |p|.
+    1. Return.
+1. If |existing state| is [=permission/denied=]:
+    1. [=Queue a global task=] on the [=permission task source=] given |global| to [=reject=] |p| with a "{{NotAllowedError}}" {{DOMException}}.
+    1. Return.
+1. Assert that |doc|'s [=node navigable=] is a [=traversable navigable=].
+1. If this algorithm was invoked when |doc|'s {{Window}} object did not have [=transient activation=]:
+    1. [=Queue a global task=] on the [=permission task source=] given |global| to [=reject=] |p| with a n "{{InvalidStateError}}" {{DOMException}}.
+    1. Return.
+1. Let |embeddee opt-in| be the result of running an [=implementation-defined=] set of steps to determine if |descriptor|'s {{TopLevelStorageAccessPermissionDescriptor/requestedOrigin}} opts into sharing on |top-level site|.
+1. If |embeddee opt-in| is false:
+    1. [=Queue a global task=] on the [=permission task source=] given |global| to [=reject=] |p| with a "{{NotAllowedError}}" {{DOMException}}.
+    1. Return.
+1. Let |implicitly granted| and |implicitly denied| (each a [=boolean=]) be the result of running an [=implementation-defined=] set of steps to determine if |top-level site|'s request for |descriptor|'s {{TopLevelStorageAccessPermissionDescriptor/requestedOrigin}} should be granted or denied without prompting the user.
+1. If |implicitly granted| is true:
+    1. [=Queue a global task=] on the [=permission task source=] given |global| to [=resolve=] |p|.
+    1. Return.
+1. If |implicitly denied| is true:
+    1. [=reject=] |p| with a "{{NotAllowedError}}" {{DOMException}}.
+    1. Return.
+1. Let |permissionState| be the result of [=requesting permission to use=] "<a permission><code>top-level-storage-access</code></a>" with |descriptor|.
+1. If |permissionState| is [=permission/granted=]:
+    1. [=Queue a global task=] on the [=permission task source=] given |global| to [=resolve=] |p|.
+    1. Return.
+1. If |doc|'s {{Window}} object has [=transient activation=], [=consume user activation=] with it.
```

Comment:
User activation querying and consumption should be done on the main thread. How does this work implementation-wise?